File: bindings/python/python/auths/gateway.py
"""Proof/action-only client for a separately operated Auths gateway.

This client does not accept a provider URL, method, body, header, or credential.
It does not establish deployment isolation: that is an operator property.
Besides submission, it can ask the gateway for one signed observation (a
read-back of the recipe's observed field, or the stored outcome of a logical
operation) to attach to a later action; an observation request never writes.
"""

from __future__ import annotations

import asyncio
import base64
import json
import re
import struct
from dataclasses import dataclass
from pathlib import Path
from typing import Final, Literal, Mapping, Optional, Union, cast

_REQUEST_SCHEMA = "auths.gateway-submit/1"
_OBSERVE_SCHEMA = "auths.gateway-observe/1"
_READ_BACK_SCHEMA: Final = "auths.gateway-readback/1"
_OUTCOME_SCHEMA: Final = "auths.gateway-outcome/1"
_OBSERVATION_MEDIA_TYPE = "application/vnd.auths.observation.v1+cbor"
_MAX_OBSERVATION_BYTES = 4_096
_MAX_SUBJECT_BYTES = 1_024
_MAX_READ_BACK_ARGUMENTS = 16
_MAX_ARGUMENT_BYTES = 128
_MAX_PROOF_BYTES = 4 * 1024 * 1024
_MAX_ACTION_BYTES = 64 * 1024
_MAX_RESPONSE_BYTES = 8 * 1024
_ECHO = re.compile(r"auths-e1-[0-9a-f]{64}")
_DIGEST = re.compile(r"[0-9a-f]{64}")
_FIELD_NAME = re.compile(r"[A-Za-z0-9][A-Za-z0-9_.-]{0,63}")
_OPERATION_ID = re.compile(r"[A-Za-z0-9][A-Za-z0-9._-]{0,127}")
_BASE64URL = re.compile(r"[A-Za-z0-9_-]*")
_MAX_U64 = 2**64 - 1


class GatewayProtocolError(RuntimeError):
    """The gateway socket or its bounded response was unavailable or malformed."""


@dataclass(frozen=True)
class GatewayEndpoint:
    """Absolute path to an operator-provisioned local application socket."""

    path: Path

    def __post_init__(self) -> None:
        if (
            not isinstance(self.path, Path)
            or not self.path.is_absolute()
            or "\x00" in str(self.path)
        ):
            raise ValueError("gateway endpoint must be an absolute Unix socket path")
        if len(str(self.path).encode()) > 100:
            raise ValueError("gateway socket path is too long")


@dataclass(frozen=True)
class GatewayDenied:
    """Native proof verification denied the exact action."""

    code: str
    outcome: Literal["denied"] = "denied"


@dataclass(frozen=True)
class GatewayIndeterminate:
    """Native proof verification could not authorize the action."""

    code: str
    outcome: Literal["indeterminate"] = "indeterminate"


@dataclass(frozen=True)
class GatewayNotEntered:
    """The gateway refused provider entry after proof verification."""

    code: str
    outcome: Literal["not-entered"] = "not-entered"


@dataclass(frozen=True)
class GatewayUnknown:
    """Provider entry or effect is ambiguous; do not automatically retry."""

    outcome: Literal["unknown"] = "unknown"


@dataclass(frozen=True)
class GatewayResponseRecorded:
    """A complete HTTP response was recorded; effect is not established."""

    status: int
    outcome: Literal["response-recorded"] = "response-recorded"


@dataclass(frozen=True)
class GatewayObserved:
    """Separate read-back matched or differed; match is not causation."""

    status: int
    matched: bool
    outcome: Literal["observed"] = "observed"


@dataclass(frozen=True)
class GatewayProviderEvidence:
    """Secret-free summary of provider-held read-back evidence.

    The response bytes and observation locator stay in the operator's attempt
    store. ``observed_at`` is gateway wall-clock seconds and is not
    authenticated.
    """

    channel: Literal["read-back"]
    echo: str
    evidence_digest: str
    observed_at: int


@dataclass(frozen=True)
class GatewayObservedByProvider:
    """Read-back returned this attempt's echo token and the verified value.

    The token is derived from the authorized action, but anyone who knows the
    namespace, operation ID, and action commitment can compute it: the link
    holds only while no other party with write access to that provider field
    wrote the same token. ``status`` is ``None`` when the write was unknown.
    """

    status: Optional[int]
    evidence: GatewayProviderEvidence
    outcome: Literal["observed-by-provider"] = "observed-by-provider"


GatewayResult = Union[
    GatewayDenied,
    GatewayIndeterminate,
    GatewayNotEntered,
    GatewayUnknown,
    GatewayResponseRecorded,
    GatewayObserved,
    GatewayObservedByProvider,
]


def _valid_status(status: object) -> bool:
    return (
        isinstance(status, int)
        and not isinstance(status, bool)
        and 100 <= status <= 599
    )


def _parse_evidence(value: object) -> GatewayProviderEvidence:
    if not isinstance(value, dict):
        raise GatewayProtocolError("gateway returned invalid evidence")
    evidence = cast(Mapping[str, object], value)
    channel = evidence.get("channel")
    echo = evidence.get("echo")
    digest = evidence.get("evidence_digest")
    observed_at = evidence.get("observed_at")
    if (
        set(evidence) != {"channel", "echo", "evidence_digest", "observed_at"}
        or channel != "read-back"
        or not isinstance(echo, str)
        or _ECHO.fullmatch(echo) is None
        or not isinstance(digest, str)
        or _DIGEST.fullmatch(digest) is None
        or not isinstance(observed_at, int)
        or isinstance(observed_at, bool)
        or not 0 <= observed_at <= _MAX_U64
    ):
        raise GatewayProtocolError("gateway returned invalid evidence")
    return GatewayProviderEvidence("read-back", echo, digest, observed_at)


def _parse_result(raw: bytes) -> GatewayResult:
    try:
        decoded: object = json.loads(raw)
    except (UnicodeDecodeError, ValueError) as error:
        raise GatewayProtocolError("gateway returned invalid JSON") from error
    if not isinstance(decoded, dict):
        raise GatewayProtocolError("gateway returned invalid result")
    value = cast(Mapping[str, object], decoded)
    outcome = value.get("outcome")
    if outcome in {"denied", "indeterminate", "not-entered"}:
        code = value.get("code")
        if set(value) != {"outcome", "code"} or not isinstance(code, str):
            raise GatewayProtocolError("gateway returned invalid refusal")
        if not 1 <= len(code) <= 128:
            raise GatewayProtocolError("gateway returned invalid code")
        if outcome == "denied":
            return GatewayDenied(code)
        if outcome == "indeterminate":
            return GatewayIndeterminate(code)
        return GatewayNotEntered(code)
    if outcome == "unknown" and set(value) == {"outcome"}:
        return GatewayUnknown()
    if outcome == "observed-by-provider":
        status = value.get("status")
        if set(value) != {"outcome", "status", "evidence"} or not (
            status is None or _valid_status(status)
        ):
            raise GatewayProtocolError("gateway returned invalid provider observation")
        return GatewayObservedByProvider(
            cast(Optional[int], status), _parse_evidence(value.get("evidence"))
        )
    if outcome in {"response-recorded", "observed"}:
        expected = (
            {"outcome", "status"}
            if outcome == "response-recorded"
            else {"outcome", "status", "matched"}
        )
        status = value.get("status")
        if set(value) != expected or not _valid_status(status):
            raise GatewayProtocolError("gateway returned invalid response stage")
        status = cast(int, status)
        if outcome == "response-recorded":
            return GatewayResponseRecorded(status)
        matched = value.get("matched")
        if not isinstance(matched, bool):
            raise GatewayProtocolError("gateway returned invalid observation")
        return GatewayObserved(status, matched)
    raise GatewayProtocolError("gateway returned unknown result stage")


@dataclass(frozen=True)
class GatewaySignedObservation:
    """One canonical gateway-signed observation to attach to a later action.

    ``observation`` is the exact signed CBOR; attach it unmodified as a
    detached attachment whose descriptor media type is ``media_type``. The
    observation asserts only what the gateway saw at ``observed_at`` (gateway
    wall-clock seconds, covered by the signature); it is not a write receipt
    and does not prove the fact still holds. ``schema`` is
    ``auths.gateway-readback/1`` for a read-back and ``auths.gateway-outcome/1``
    for an operation outcome.
    """

    schema: Literal["auths.gateway-readback/1", "auths.gateway-outcome/1"]
    subject: str
    observed_at: int
    media_type: Literal["application/vnd.auths.observation.v1+cbor"]
    observation: bytes
    outcome: Literal["signed"] = "signed"


@dataclass(frozen=True)
class GatewayObservationRefused:
    """The gateway signed nothing; ``code`` is its stable refusal code."""

    code: str
    outcome: Literal["refused"] = "refused"


GatewayObserveResult = Union[GatewaySignedObservation, GatewayObservationRefused]


def _encode_base64url(value: bytes) -> str:
    return base64.urlsafe_b64encode(value).rstrip(b"=").decode("ascii")


def _decode_base64url(text: str, maximum: int) -> bytes:
    """Decode canonical unpadded base64url, rejecting any other spelling."""
    if (
        len(text) > (maximum * 4 + 2) // 3
        or _BASE64URL.fullmatch(text) is None
        or len(text) % 4 == 1
    ):
        raise GatewayProtocolError("gateway returned invalid observation encoding")
    decoded = base64.urlsafe_b64decode(text + "=" * (-len(text) % 4))
    if not 1 <= len(decoded) <= maximum or _encode_base64url(decoded) != text:
        raise GatewayProtocolError("gateway returned invalid observation encoding")
    return decoded


def _parse_observe_result(
    raw: bytes, schema: Literal["auths.gateway-readback/1", "auths.gateway-outcome/1"]
) -> GatewayObserveResult:
    try:
        decoded: object = json.loads(raw)
    except (UnicodeDecodeError, ValueError) as error:
        raise GatewayProtocolError("gateway returned invalid JSON") from error
    if not isinstance(decoded, dict):
        raise GatewayProtocolError("gateway returned invalid observation result")
    value = cast(Mapping[str, object], decoded)
    outcome = value.get("outcome")
    if outcome == "refused":
        code = value.get("code")
        if (
            set(value) != {"outcome", "code"}
            or not isinstance(code, str)
            or not 1 <= len(code) <= 128
        ):
            raise GatewayProtocolError("gateway returned invalid refusal")
        return GatewayObservationRefused(code)
    if outcome != "signed":
        raise GatewayProtocolError("gateway returned unknown observation result")
    subject = value.get("subject")
    observed_at = value.get("observed_at")
    encoded = value.get("observation_b64")
    if (
        set(value)
        != {
            "outcome",
            "schema",
            "subject",
            "observed_at",
            "media_type",
            "observation_b64",
        }
        or value.get("schema") != schema
        or value.get("media_type") != _OBSERVATION_MEDIA_TYPE
        or not isinstance(subject, str)
        or not 1 <= len(subject.encode("utf-8", "surrogatepass")) <= _MAX_SUBJECT_BYTES
        or not isinstance(observed_at, int)
        or isinstance(observed_at, bool)
        or not 0 <= observed_at <= _MAX_U64
        or not isinstance(encoded, str)
    ):
        raise GatewayProtocolError("gateway returned invalid signed observation")
    return GatewaySignedObservation(
        schema,
        subject,
        observed_at,
        "application/vnd.auths.observation.v1+cbor",
        _decode_base64url(encoded, _MAX_OBSERVATION_BYTES),
    )


def _read_back_request(arguments: Mapping[str, str]) -> dict[str, object]:
    if not isinstance(arguments, Mapping) or not (
        1 <= len(arguments) <= _MAX_READ_BACK_ARGUMENTS
    ):
        raise ValueError("read-back arguments must be a bounded mapping")
    checked: dict[str, str] = {}
    for name, value in cast(Mapping[object, object], arguments).items():
        if not isinstance(name, str) or _FIELD_NAME.fullmatch(name) is None:
            raise ValueError("read-back argument names must be recipe field names")
        if (
            not isinstance(value, str)
            or "\x00" in value
            or not 1
            <= len(value.encode("utf-8", "surrogatepass"))
            <= _MAX_ARGUMENT_BYTES
        ):
            raise ValueError("read-back argument values must be bounded strings")
        checked[name] = value
    return {"kind": "read-back", "arguments": checked}


class GatewayClient:
    """Submit only canonical proof and action bytes over the local app socket,
    or request one read-only gateway-signed observation."""

    def __init__(self, endpoint: GatewayEndpoint) -> None:
        if not isinstance(endpoint, GatewayEndpoint):
            raise TypeError("endpoint must be a GatewayEndpoint")
        self._endpoint = endpoint

    async def submit(self, *, proof: bytes, action: bytes) -> GatewayResult:
        """Return proof/transport stage without implying provider effect.

        An IPC failure is not a retry signal: the gateway may have entered the
        provider before the client lost its response.
        """
        if not isinstance(proof, bytes) or not 1 <= len(proof) <= _MAX_PROOF_BYTES:
            raise ValueError("proof must be bounded bytes")
        if not isinstance(action, bytes) or not 1 <= len(action) <= _MAX_ACTION_BYTES:
            raise ValueError("action must be bounded bytes")
        payload = json.dumps(
            {
                "schema": _REQUEST_SCHEMA,
                "proof_b64": _encode_base64url(proof),
                "action_b64": _encode_base64url(action),
            },
            separators=(",", ":"),
        ).encode("utf-8")
        return _parse_result(
            await self._exchange(
                payload, "gateway exchange unavailable; outcome may be unknown"
            )
        )

    async def observe_read_back(
        self, arguments: Mapping[str, str]
    ) -> GatewayObserveResult:
        """Ask the gateway to read the recipe's observed field now and sign it.

        ``arguments`` names exactly the recipe observation path's fields (for
        example ``{"record_id": "rec..."}``); the gateway derives the URL,
        subject, and time itself. The application learns the observed value.

        Raises ``ValueError`` for unbounded or non-string arguments and
        ``GatewayProtocolError`` when the socket or bounded response is
        unavailable or malformed. A refusal is a result, not an exception.
        """
        request = _read_back_request(arguments)
        return await self._observe(request, _READ_BACK_SCHEMA)

    async def observe_outcome(self, operation_id: str) -> GatewayObserveResult:
        """Ask the gateway to sign the stored outcome of one logical operation.

        The gateway reads only its own attempt store; no provider is contacted.
        ``operation_id`` is the 1-128-byte canonical ASCII token used when the
        operation was submitted.

        Raises ``ValueError`` for a noncanonical operation ID and
        ``GatewayProtocolError`` for an unavailable or malformed exchange.
        """
        if (
            not isinstance(operation_id, str)
            or _OPERATION_ID.fullmatch(operation_id) is None
        ):
            raise ValueError("operation ID must be a canonical 1-128-byte token")
        return await self._observe(
            {"kind": "outcome", "operation_id": operation_id}, _OUTCOME_SCHEMA
        )

    async def _observe(
        self,
        request: Mapping[str, object],
        schema: Literal["auths.gateway-readback/1", "auths.gateway-outcome/1"],
    ) -> GatewayObserveResult:
        payload = json.dumps(
            {"schema": _OBSERVE_SCHEMA, "request": request},
            separators=(",", ":"),
            ensure_ascii=False,
        ).encode("utf-8")
        return _parse_observe_result(
            await self._exchange(payload, "gateway observation exchange unavailable"),
            schema,
        )

    async def _exchange(self, payload: bytes, failure: str) -> bytes:
        try:
            reader, writer = await asyncio.wait_for(
                asyncio.open_unix_connection(str(self._endpoint.path)), timeout=5
            )
            try:
                writer.write(struct.pack(">I", len(payload)) + payload)
                await asyncio.wait_for(writer.drain(), timeout=5)
                length = struct.unpack(
                    ">I", await asyncio.wait_for(reader.readexactly(4), timeout=45)
                )[0]
                if not 1 <= length <= _MAX_RESPONSE_BYTES:
                    raise GatewayProtocolError("gateway response exceeded the bound")
                response = await asyncio.wait_for(reader.readexactly(length), timeout=5)
            finally:
                writer.close()
                await writer.wait_closed()
        except (OSError, asyncio.IncompleteReadError, TimeoutError) as error:
            raise GatewayProtocolError(failure) from error
        return response


__all__ = [
    "GatewayClient",
    "GatewayDenied",
    "GatewayEndpoint",
    "GatewayIndeterminate",
    "GatewayNotEntered",
    "GatewayObservationRefused",
    "GatewayObserveResult",
    "GatewayObserved",
    "GatewayObservedByProvider",
    "GatewayProtocolError",
    "GatewayProviderEvidence",
    "GatewayResponseRecorded",
    "GatewayResult",
    "GatewaySignedObservation",
    "GatewayUnknown",
]

File: bindings/typescript/src/gateway.ts
/**
 * Proof/action-only client for a separately operated local Auths gateway. It can also
 * ask the gateway for one read-only signed observation to attach to a later action.
 */

import { createConnection } from "node:net";

const REQUEST_SCHEMA = "auths.gateway-submit/1";
const OBSERVE_SCHEMA = "auths.gateway-observe/1";
const READ_BACK_SCHEMA = "auths.gateway-readback/1";
const OUTCOME_SCHEMA = "auths.gateway-outcome/1";
const OBSERVATION_MEDIA_TYPE = "application/vnd.auths.observation.v1+cbor";
const MAX_OBSERVATION_BYTES = 4_096;
const MAX_SUBJECT_BYTES = 1_024;
const MAX_READ_BACK_ARGUMENTS = 16;
const MAX_ARGUMENT_BYTES = 128;
const MAX_PROOF_BYTES = 4 * 1024 * 1024;
const MAX_ACTION_BYTES = 64 * 1024;
const MAX_RESPONSE_BYTES = 8 * 1024;
const MAX_FRAME_BYTES = 8 * 1024 * 1024;
const ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";
const ECHO = /^auths-e1-[0-9a-f]{64}$/;
const DIGEST = /^[0-9a-f]{64}$/;
const FIELD_NAME = /^[A-Za-z0-9][A-Za-z0-9_.-]{0,63}$/;
const OPERATION_ID = /^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$/;
const BASE64URL = /^[A-Za-z0-9_-]*$/;

type ObservationSchema = typeof READ_BACK_SCHEMA | typeof OUTCOME_SCHEMA;

/** A socket or bounded gateway response could not be trusted. Never retry a write automatically. */
export class GatewayProtocolError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "GatewayProtocolError";
  }
}

/** Operator-provisioned local application socket, never a provider URL. */
export class GatewayEndpoint {
  readonly path: string;

  constructor(path: string) {
    if (typeof path !== "string" || !path.startsWith("/") || path.includes("\0") || new TextEncoder().encode(path).length > 100) {
      throw new TypeError("gateway endpoint must be a short absolute Unix socket path");
    }
    this.path = path;
  }
}

/**
 * Secret-free summary of provider-held read-back evidence. The response bytes and
 * observation locator stay in the operator's attempt store; `observedAt` is gateway
 * wall-clock seconds and is not authenticated.
 */
export type GatewayProviderEvidence = Readonly<{
  channel: "read-back";
  echo: string;
  evidenceDigest: string;
  observedAt: number;
}>;

/**
 * Proof decision and transport/observation stages, without effect qualification.
 * `observed-by-provider` means a read-back returned this attempt's echo token and the
 * verified value; anyone who knows the namespace, operation ID, and action commitment
 * can compute that token, so the link holds only while no other writer of that
 * provider field wrote the same token. Its `status` is `null` when the write was unknown.
 */
export type GatewayResult =
  | Readonly<{ outcome: "denied"; code: string }>
  | Readonly<{ outcome: "indeterminate"; code: string }>
  | Readonly<{ outcome: "not-entered"; code: string }>
  | Readonly<{ outcome: "unknown" }>
  | Readonly<{ outcome: "response-recorded"; status: number }>
  | Readonly<{ outcome: "observed"; status: number; matched: boolean }>
  | Readonly<{ outcome: "observed-by-provider"; status: number | null; evidence: GatewayProviderEvidence }>;

/**
 * One canonical gateway-signed observation. Attach `observation` unmodified as a
 * detached attachment whose descriptor media type is `mediaType`. It asserts only what
 * the gateway saw at `observedAt` (gateway wall-clock seconds, covered by the
 * signature); it is not a write receipt and does not prove the fact still holds.
 */
export type GatewaySignedObservation = Readonly<{
  outcome: "signed";
  schema: "auths.gateway-readback/1" | "auths.gateway-outcome/1";
  subject: string;
  observedAt: number;
  mediaType: "application/vnd.auths.observation.v1+cbor";
  observation: Uint8Array;
}>;

/** Closed result of an observation request: signed bytes, or a refusal that signed nothing. */
export type GatewayObserveResult =
  | GatewaySignedObservation
  | Readonly<{ outcome: "refused"; code: string }>;

function encodeBase64Url(bytes: Uint8Array): string {
  let encoded = "";
  for (let index = 0; index < bytes.length; index += 3) {
    const first = bytes[index]!;
    const second = bytes[index + 1];
    const third = bytes[index + 2];
    encoded += ALPHABET[first >> 2]!;
    encoded += ALPHABET[((first & 3) << 4) | ((second ?? 0) >> 4)]!;
    if (second !== undefined) {
      encoded += ALPHABET[((second & 15) << 2) | ((third ?? 0) >> 6)]!;
    }
    if (third !== undefined) {
      encoded += ALPHABET[third & 63]!;
    }
  }
  return encoded;
}

/** Decodes canonical unpadded base64url only; any other spelling is rejected. */
function decodeBase64Url(text: string, maximum: number): Uint8Array {
  if (text.length > Math.ceil((maximum * 4) / 3) || !BASE64URL.test(text) || text.length % 4 === 1) {
    throw new GatewayProtocolError("gateway returned invalid observation encoding");
  }
  const bytes = new Uint8Array(Math.floor((text.length * 3) / 4));
  let buffer = 0;
  let bits = 0;
  let offset = 0;
  for (const character of text) {
    buffer = (buffer << 6) | ALPHABET.indexOf(character);
    bits += 6;
    if (bits >= 8) {
      bits -= 8;
      bytes[offset] = (buffer >> bits) & 0xff;
      offset += 1;
    }
    buffer &= (1 << bits) - 1;
  }
  if (buffer !== 0 || bytes.length < 1 || bytes.length > maximum) {
    throw new GatewayProtocolError("gateway returned invalid observation encoding");
  }
  return bytes;
}

function utf8Length(value: string): number {
  return new TextEncoder().encode(value).length;
}

function exactKeys(value: Record<string, unknown>, expected: readonly string[]): boolean {
  const actual = Object.keys(value);
  return actual.length === expected.length && expected.every((key) => Object.hasOwn(value, key));
}

function validStatus(status: unknown): status is number {
  return typeof status === "number" && Number.isInteger(status) && status >= 100 && status <= 599;
}

function parseEvidence(value: unknown): GatewayProviderEvidence {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new GatewayProtocolError("gateway returned invalid evidence");
  }
  const evidence = value as Record<string, unknown>;
  const { channel, echo, evidence_digest: digest, observed_at: observedAt } = evidence;
  if (
    !exactKeys(evidence, ["channel", "echo", "evidence_digest", "observed_at"])
    || channel !== "read-back"
    || typeof echo !== "string" || !ECHO.test(echo)
    || typeof digest !== "string" || !DIGEST.test(digest)
    || typeof observedAt !== "number" || !Number.isSafeInteger(observedAt) || observedAt < 0
  ) {
    throw new GatewayProtocolError("gateway returned invalid evidence");
  }
  return { channel, echo, evidenceDigest: digest, observedAt };
}

function parseResult(bytes: Uint8Array): GatewayResult {
  let decoded: unknown;
  try {
    decoded = JSON.parse(new TextDecoder("utf-8", { fatal: true }).decode(bytes));
  } catch {
    throw new GatewayProtocolError("gateway returned invalid JSON");
  }
  if (typeof decoded !== "object" || decoded === null || Array.isArray(decoded)) {
    throw new GatewayProtocolError("gateway returned invalid result");
  }
  const value = decoded as Record<string, unknown>;
  const outcome = value.outcome;
  if (outcome === "denied" || outcome === "indeterminate" || outcome === "not-entered") {
    if (!exactKeys(value, ["outcome", "code"]) || typeof value.code !== "string" || value.code.length < 1 || value.code.length > 128) {
      throw new GatewayProtocolError("gateway returned invalid refusal");
    }
    return { outcome, code: value.code };
  }
  if (outcome === "unknown" && exactKeys(value, ["outcome"])) {
    return { outcome };
  }
  if (outcome === "observed-by-provider") {
    if (!exactKeys(value, ["outcome", "status", "evidence"]) || !(value.status === null || validStatus(value.status))) {
      throw new GatewayProtocolError("gateway returned invalid provider observation");
    }
    return { outcome, status: value.status, evidence: parseEvidence(value.evidence) };
  }
  if (outcome === "response-recorded" || outcome === "observed") {
    const expected = outcome === "observed" ? ["outcome", "status", "matched"] : ["outcome", "status"];
    if (!exactKeys(value, expected) || !validStatus(value.status)) {
      throw new GatewayProtocolError("gateway returned invalid response stage");
    }
    if (outcome === "response-recorded") {
      return { outcome, status: value.status };
    }
    if (typeof value.matched !== "boolean") {
      throw new GatewayProtocolError("gateway returned invalid observation");
    }
    return { outcome, status: value.status, matched: value.matched };
  }
  throw new GatewayProtocolError("gateway returned unknown result stage");
}

function parseObserveResult(bytes: Uint8Array, schema: ObservationSchema): GatewayObserveResult {
  let decoded: unknown;
  try {
    decoded = JSON.parse(new TextDecoder("utf-8", { fatal: true }).decode(bytes));
  } catch {
    throw new GatewayProtocolError("gateway returned invalid JSON");
  }
  if (typeof decoded !== "object" || decoded === null || Array.isArray(decoded)) {
    throw new GatewayProtocolError("gateway returned invalid observation result");
  }
  const value = decoded as Record<string, unknown>;
  if (value.outcome === "refused") {
    if (!exactKeys(value, ["outcome", "code"]) || typeof value.code !== "string" || value.code.length < 1 || value.code.length > 128) {
      throw new GatewayProtocolError("gateway returned invalid refusal");
    }
    return { outcome: "refused", code: value.code };
  }
  if (value.outcome !== "signed") {
    throw new GatewayProtocolError("gateway returned unknown observation result");
  }
  const { subject, observed_at: observedAt, observation_b64: encoded } = value;
  if (
    !exactKeys(value, ["outcome", "schema", "subject", "observed_at", "media_type", "observation_b64"])
    || value.schema !== schema
    || value.media_type !== OBSERVATION_MEDIA_TYPE
    || typeof subject !== "string" || subject.length < 1 || utf8Length(subject) > MAX_SUBJECT_BYTES
    || typeof observedAt !== "number" || !Number.isSafeInteger(observedAt) || observedAt < 0
    || typeof encoded !== "string"
  ) {
    throw new GatewayProtocolError("gateway returned invalid signed observation");
  }
  return {
    outcome: "signed",
    schema,
    subject,
    observedAt,
    mediaType: OBSERVATION_MEDIA_TYPE,
    observation: decodeBase64Url(encoded, MAX_OBSERVATION_BYTES),
  };
}

function readBackRequest(argumentsMap: Readonly<Record<string, string>>): Record<string, unknown> {
  if (typeof argumentsMap !== "object" || argumentsMap === null || Array.isArray(argumentsMap)) {
    throw new TypeError("read-back arguments must be a bounded object");
  }
  const entries = Object.entries(argumentsMap as Record<string, unknown>);
  if (entries.length < 1 || entries.length > MAX_READ_BACK_ARGUMENTS) {
    throw new TypeError("read-back arguments must be a bounded object");
  }
  const checked: Record<string, string> = {};
  for (const [name, value] of entries) {
    if (!FIELD_NAME.test(name)) {
      throw new TypeError("read-back argument names must be recipe field names");
    }
    if (typeof value !== "string" || value.includes("\0") || value.length < 1 || utf8Length(value) > MAX_ARGUMENT_BYTES) {
      throw new TypeError("read-back argument values must be bounded strings");
    }
    checked[name] = value;
  }
  return { kind: "read-back", arguments: checked };
}

function concatenate(parts: readonly Uint8Array[], total: number): Uint8Array {
  const bytes = new Uint8Array(total);
  let offset = 0;
  for (const part of parts) {
    bytes.set(part, offset);
    offset += part.length;
  }
  return bytes;
}

/**
 * Submit only canonical proof and action bytes, or request one read-only signed
 * observation; no credential or raw HTTP API exists.
 */
export class GatewayClient {
  readonly #endpoint: GatewayEndpoint;

  constructor(endpoint: GatewayEndpoint) {
    if (!(endpoint instanceof GatewayEndpoint)) {
      throw new TypeError("endpoint must be a GatewayEndpoint");
    }
    this.#endpoint = endpoint;
  }

  async submit(input: Readonly<{ proof: Uint8Array; action: Uint8Array }>): Promise<GatewayResult> {
    if (!(input.proof instanceof Uint8Array) || input.proof.length < 1 || input.proof.length > MAX_PROOF_BYTES) {
      throw new TypeError("proof must be bounded bytes");
    }
    if (!(input.action instanceof Uint8Array) || input.action.length < 1 || input.action.length > MAX_ACTION_BYTES) {
      throw new TypeError("action must be bounded bytes");
    }
    const payload = new TextEncoder().encode(JSON.stringify({
      schema: REQUEST_SCHEMA,
      proof_b64: encodeBase64Url(input.proof),
      action_b64: encodeBase64Url(input.action),
    }));
    return this.#exchange(payload, "outcome may be unknown", parseResult);
  }

  /**
   * Asks the gateway to read the recipe's observed field now and sign what it read.
   * `argumentsMap` names exactly the recipe observation path's fields (for example
   * `{ record_id: "rec..." }`); the gateway derives URL, subject, and time itself, and
   * the application learns the observed value. Throws `TypeError` for unbounded or
   * non-string arguments and `GatewayProtocolError` for an unavailable or malformed
   * exchange; a refusal is a result, not an exception.
   */
  async observeReadBack(argumentsMap: Readonly<Record<string, string>>): Promise<GatewayObserveResult> {
    return this.#observe(readBackRequest(argumentsMap), READ_BACK_SCHEMA);
  }

  /**
   * Asks the gateway to sign the stored outcome of one logical operation. The gateway
   * reads only its own attempt store; no provider is contacted. Throws `TypeError` for a
   * noncanonical 1-128-byte operation ID.
   */
  async observeOutcome(operationId: string): Promise<GatewayObserveResult> {
    if (typeof operationId !== "string" || !OPERATION_ID.test(operationId)) {
      throw new TypeError("operation ID must be a canonical 1-128-byte token");
    }
    return this.#observe({ kind: "outcome", operation_id: operationId }, OUTCOME_SCHEMA);
  }

  #observe(request: Record<string, unknown>, schema: ObservationSchema): Promise<GatewayObserveResult> {
    const payload = new TextEncoder().encode(JSON.stringify({ schema: OBSERVE_SCHEMA, request }));
    return this.#exchange(payload, "observation not obtained", (bytes) => parseObserveResult(bytes, schema));
  }

  #exchange<T>(payload: Uint8Array, consequence: string, parse: (bytes: Uint8Array) => T): Promise<T> {
    if (payload.length > MAX_FRAME_BYTES) {
      throw new TypeError("gateway request exceeds frame bound");
    }
    const frame = new Uint8Array(payload.length + 4);
    new DataView(frame.buffer).setUint32(0, payload.length, false);
    frame.set(payload, 4);
    return new Promise<T>((resolve, reject) => {
      const socket = createConnection({ path: this.#endpoint.path });
      const parts: Uint8Array[] = [];
      let received = 0;
      let settled = false;
      const finish = (result: { value: T } | GatewayProtocolError): void => {
        if (settled) return;
        settled = true;
        clearTimeout(timer);
        socket.destroy();
        if (result instanceof GatewayProtocolError) reject(result);
        else resolve(result.value);
      };
      const timer = setTimeout(() => finish(new GatewayProtocolError(`gateway exchange timed out; ${consequence}`)), 45_000);
      socket.once("connect", () => { socket.write(frame); });
      socket.on("data", (chunk) => {
        parts.push(chunk);
        received += chunk.length;
        if (received > MAX_RESPONSE_BYTES + 4) {
          finish(new GatewayProtocolError("gateway response exceeded the bound"));
          return;
        }
        if (received < 4) return;
        const bytes = concatenate(parts, received);
        const length = new DataView(bytes.buffer).getUint32(0, false);
        if (length < 1 || length > MAX_RESPONSE_BYTES) {
          finish(new GatewayProtocolError("gateway response exceeded the bound"));
        } else if (received >= length + 4) {
          try {
            finish({ value: parse(bytes.subarray(4, length + 4)) });
          } catch {
            finish(new GatewayProtocolError(`gateway returned invalid response; ${consequence}`));
          }
        }
      });
      socket.once("error", () => finish(new GatewayProtocolError(`gateway exchange unavailable; ${consequence}`)));
      socket.once("end", () => finish(new GatewayProtocolError(`gateway response incomplete; ${consequence}`)));
    });
  }
}
